### DISCUSSION TITLE
About NeighborhoodGenerators

### DISCUSSION THREAD

**Discussion by C8504** [google/or-tools#2943] [CP-SAT questions] [answered]
Hi Experts,

I am reading NeighborhoodGenerators code, while it is so complex to understand, so I am wondering if have any papers about this.

Such as:
RelaxRandomVariablesGenerator/RelaxRandomConstraintsGenerator/VariableGraphNeighborhoodGenerator/ConstraintGraphNeighborhoodGenerator

Thanks a lot!
John

**Comment by lperron** [comment, marked as answer]
There are no papers. There are not that complex. 
for the graph ones: we use the variable constraint graph, we start traversing it, and relax all variables we encounter.
for the random ones: we relax a random subset of variables, or pick a random subset of constraints, and relax all variables in it.
